### ISSUE TITLE
[SR-16071] Substring hashing broken by out-of-range storage character in Swift 5.6

### ISSUE THREAD

**Issue by stevebrambilla** [apple/swift#58332]
|                  |                 |
   |------------------|-----------------|
   |Previous ID       | SR-16071      |
   |Radar             | None         |
   |Original Reporter | @stevebrambilla      |
   |Type              | Bug    |

<details>
  <summary>Environment</summary>

Xcode 13.3  
Swift 5.6

</details>

   <details>
  <summary>Additional Detail from JIRA</summary>

   |                  |                 |
   |------------------|-----------------|
   |Votes             | 0         |
   |Component/s       |     |
   |Labels            | Bug        |
   |Assignee          | None      |
   |Priority          | Medium      |

   

   md5: 7794959d7c2df469a1658020927c65d2

  </details>

**Issue Description:**

The test below shows an example of two substrings that are equal, but their hash values aren't.

This is triggered by the "\u{2019}" character in the second string. If you remove that character then hashing behaves as expected.

 

``` java
func testSubstringHashing() {
    let firstLines: [Substring] = """
        Line1
        Line2: nil
        """
        .split(separator: "\n")
    
    // The \u{2019} character triggers a hash difference when comparing the first line
    let secondLines: [Substring] = """
        Line1
        Line2: "\u{2019}"
        """
        .split(separator: "\n")
    
    var firstHasher = Hasher()
    firstHasher.combine(firstLines[0]) // "Line1"
    let firstHash = firstHasher.finalize()
    
    var secondHasher = Hasher()
    secondHasher.combine(secondLines[0]) // "Line1"
    let secondHash = secondHasher.finalize()
    
    XCTAssert(firstLines[0] == secondLines[0]) // "Line1" == "Line1"
    XCTAssert(firstHash == secondHash) // Fails
}
```

**Comment by glessard** [comment]
@stevebrambilla I believe this is fixed in Swift 5.7.1. Can you verify?

**Comment by stevebrambilla** [comment]
Yes, I can confirm that this is fixed now with 5.7.1. Thank you!
